src/interrupt/mod.rs

```diff
@@ -0,0 +1,289 @@
+///
+/// The usage of the resulted notifier object is speciffic to the hypervisor but the semantics of
+/// the object returned by the `trigger_notifier()` method should follow the semantics from
+/// `EdgeInterrupt::trigger()` or `LevelInterrupt::assert()` (e.g. when the user changes the state
+/// of the notifier object, an interrupt is queued for the guest).
+///
+/// A notable example is VFIO that allows a device to register the irqfd so that interrupts follow
+/// a fast path that doesn't require going through the VMM. Another example is XEN evtchn.
+///
+/// Implementations of this trait must provide the trigger notifier object.
+pub trait AsRefTriggerNotifier {
+    /// The type of the underlying mechanism used for trigger notifications by this interrupt.
+    type NotifierType;
+
+    /// Returns a reference to a trigger notifier from this interrupt.
+    ///
+    /// An interrupt notifier allows for external components and processes to inject interrupts
+    /// into a guest through a different interface other than `EdgeInterrupt::trigger()`.
+    fn trigger_notifier(&self) -> &Self::NotifierType;
+}
+
+/// Trait that provides access to the underlying resample notification object used by
+/// the hypervisor.
+///
+/// This enables use cases where the notification that the interrupt was resampled is
+/// handled by a component that understands the underlying hypervisor interrupt implementation
+/// and wants to bypass the VMM.
+///
+/// The semantics of the object returned by `resample_notifier()` are similar to those of
+/// `AutoRetriggerInterrupt` (when the state of the notifier object changes it means that
+/// the interrupt was resampled and the device should reassert the interrupt).
+///
+/// VFIO supports the registration of a `resamplefd` which would be returned by
+/// `resample_notifier`.
+///
+/// Implementations of this trait must provide the resample notifier object.
+pub trait AsRefResampleNotifier {
+    /// The type of the underlying mechanism used for resample notifications by an interrupt.
+    type NotifierType;
+
+    /// Returns a reference to a resample notifier from an interrupt.
+    ///
+    /// An end-of-interrupt notifier allows for external components and processes to be notified
+    /// when a guest acknowledges an interrupt. This can be used to resample and inject a
+    /// level-triggered interrupt, or to mitigate the effect of lost timer interrupts.
+    fn resample_notifier(&self) -> &Self::NotifierType;
+}
+
+/// Trait to manage a group of interrupt sources for a device.
+///
+/// A device may use an InterruptSourceGroup to manage multiple interrupts of the same type.
+/// The group allows a device to request and release interrupts and perform actions on the
+/// whole collection of interrupts like enable and disable for cases where enabling or disabling
+/// a single interrupt in the group does not make sense. For example, PCI MSI interrupts must be
+/// enabled as a group.
```

Comment:
Is this a KVM limitation or something that comes as a requirement from PCI passthrough?

Reply:
MSI or MSI-x is enabled by setting the enable bit in the message control struct. This refers to the whole set of MSI interrupts. You can mask each individual interrupt but you can't enable/disable them individually.

Reply:
Is there a functional difference between masked and disabled MSI-X vectors from the interrupt subsystems perspective?

Reply:
Some controllers may have different registers to distinguish masked vs enabled interrupts. Take a look here: https://docs.microsoft.com/en-us/windows-hardware/drivers/gpio/gpio-interrupt-masks

There may also be functional differences in what happens when a masked interrupt is asserted. In some cases interrupt is not delivered when asserted until it is unmasked, whereas asserting a disabled interrupt should not keep the pending state.

Reply:
This document describes behavior of a (GPIO) interrupt _controller_. So what you are saying is that the implementation of this trait needs to know what kind of interrupt controller it is connected to. This is fine, I'm just trying to wrap my head around things. :)